godot-core/src/builtin/quaternion.rs

```diff
@@ -130,70 +138,115 @@ impl Quaternion {
         Quaternion::new(v.x, v.y, v.z, 0.0)
     }
 
+    /// Normalizes the quaternion.
+    ///
+    /// # Panics
+    /// If the quaternion has length of 0.
     pub fn normalized(self) -> Self {
-        self / self.length()
+        let length = self.length();
+        assert!(!length.approx_eq(&0.0), "Quaternion has length 0");
+        self / length
     }
 
+    /// Performs slerp
+    ///
+    /// # Panics
+    /// If either quaternion is not normalized
     pub fn slerp(self, to: Self, weight: real) -> Self {
-        let mut cosom = self.dot(to);
-        let to1: Self;
-        let omega: real;
-        let sinom: real;
-        let scale0: real;
-        let scale1: real;
-        if cosom < 0.0 {
-            cosom = -cosom;
-            to1 = -to;
-        } else {
-            to1 = to;
-        }
+        let interpolated = self.as_inner().slerp(to, weight.as_f64());
 
-        if 1.0 - cosom > real::CMP_EPSILON {
-            omega = cosom.acos();
-            sinom = omega.sin();
-            scale0 = ((1.0 - weight) * omega).sin() / sinom;
-            scale1 = (weight * omega).sin() / sinom;
-        } else {
-            scale0 = 1.0 - weight;
-            scale1 = weight;
-        }
+        let error = interpolated.is_default() && (!self.is_normalized() || !to.is_normalized());
+        assert!(!error, "Slerp requires normalized quaternions");
 
-        scale0 * self + scale1 * to1
+        interpolated
     }
 
+    /// Performs slerpni
+    ///
+    /// # Panics
+    /// If either quaternion is not normalized
     pub fn slerpni(self, to: Self, weight: real) -> Self {
```

Comment:
Same story as slerp. Abandoned because it did not appropriately error on non-normalized vectors.